squid-forward-proxy-operator/src/http_proxy.py

```diff
@@ -0,0 +1,882 @@
+    auth: Tuple[str, ...]
+    src_ips: Tuple[str, ...] = tuple()
+
+    @staticmethod
+    def parse_domain(domain: str) -> Tuple[str, int]:
+        """Parse a domain string in the form of host[:port].
+
+        Args:
+            domain: The domain string.
+
+        Returns:
+            A (host, port) tuple. Port is 0 if not specified.
+
+        Raises:
+            ValueError: If the domain string is invalid.
+        """
+        # ipv6 (the correct way), i.e. "[::1]:8080" or "[::1]"
+        if domain.startswith("["):
+            search = re.findall(r"^\[([a-f0-9:A-F]+)](:[0-9]+)?$", domain)
+            if not search:
+                raise ValueError(f"invalid domain: {domain}")
+            host, port = search[0]
+            ipaddress.ip_network(host, strict=False)
+            host = f"[{host}]"
+            port = port.removeprefix(":")
+        # ipv6 (the "incorrect" way), i.e. "fe80::1", "::1"
+        elif domain.count(":") >= 2:
+            ipaddress.ip_network(domain, strict=False)
+            host, port = f"[{domain}]", 0
+        # ipv4
+        elif re.match("^[0-9.:]+$", domain):
+            if ":" in domain:
+                host, port = domain.rsplit(":", 1)
+            else:
+                host, port = domain, 0
+            ipaddress.ip_address(host)
+        # DNS domain
+        else:
+            match = re.match(
+                r"^(([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]*[a-zA-Z0-9])\.)*"
+                r"([A-Za-z0-9]|[A-Za-z0-9][A-Za-z0-9\-]*[A-Za-z0-9])"
+                r"(:[0-9]+)?$",
+                domain,
+            )
+            if not match:
+                raise ValueError(f"invalid domain: {domain}")
+            if ":" in domain:
+                host, port = domain.rsplit(":", 1)
+            else:
+                host, port = domain, 0
+        return host, int(port)
+
+    @field_validator("domains", mode="before")
+    @classmethod
+    def _validate_and_transform_domains(cls, domains: Sequence[str]) -> Tuple[str, ...]:
+        """Validate and transform the domains input.
+
+        Args:
+            domains: The input domains.
+
+        Returns:
+            The canonical representation of the domains.
+        """
+        valid_domains = []
+        invalid_domains = []
+        for domain in domains:
+            try:
+                host, port = cls.parse_domain(domain)
+                if not 0 <= port <= 65535:
+                    raise ValueError(f"invalid port: {port}")
+                if port != 0:
+                    valid_domains.append(f"{host}:{port}")
+                else:
+                    valid_domains.append(f"{host}:80")
+                    valid_domains.append(f"{host}:443")
+            except ValueError:
+                invalid_domains.append(domain)
+        if invalid_domains:
+            raise ValueError(f"invalid domains: {invalid_domains}")
+        if not valid_domains:
+            raise ValueError("no domains specified")
```

Comment:
Should be checked before the for loop as it's equivalent to no domains is specified

Reply:
Updated, thanks!